core/dbt/clients/agate_helper.py

```diff
@@ -84,21 +93,36 @@ def table_from_data(data, column_names: Iterable[str]) -> agate.Table:
         table = agate.Table.from_object(data, column_types=DEFAULT_TYPE_TESTER)
         return table.select(column_names)
 
-
 def table_from_data_flat(data, column_names: Iterable[str]) -> agate.Table:
-    "Convert list of dictionaries into an Agate table"
+    """
+    Convert a list of dictionaries into an Agate table. This method does not
+    coerce string values into more specific types (eg. '005' will not be
+    coerced to '5'). Additionally, this method does not coerce values to
+    None (eg. '' or 'null' will retain their string literal representations).
+    """
 
     rows = []
+    text_only_columns = set()
     for _row in data:
         row = []
-        for value in list(_row.values()):
+        for col_name in column_names:
```

Comment:
I think our existing implementation assumed that the order of keys in a dictionary was guaranteed... but IIRC that's not true for all versions of Python. My Python is rusty, so maybe this change is not necessary. I can revert if need be :)

Reply:
Not my area, but I do remember hearing that this changed in recent versions of python. Indeed [this blog post](https://softwaremaniacs.org/blog/2020/02/05/dicts-ordered/) suggests that dictionaries are ordered by insertion as of python 3.6, and that this is officially guaranteed as of python 3.7. dbt currently requires py36 or higher.